### ISSUE TITLE
Bug: [no-unnecessary-condition] does not report unnecessary default value during object destructuring

### ISSUE THREAD

**Issue by TkDodo** [typescript-eslint/typescript-eslint#10082]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=GYVwdgxgLglg9mABAYTgWwA4IKZigCgG9Fg45EBeRARkQF8AuRY0uJsENAI2wCd6AlMwBQiRL2xQQvJKwDcwusKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function Component({ foo = 1 }: { foo: number }) {
  return foo;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I would expect a lint error here form `no-unnecessary-condition` because the default value assignment can never be reached, given that `foo` is mandatory in the object

### Actual Result

no error shown

### Additional Info

The rule correctly reports an error when nullish coalesce is used instead:

```
function Component2(props: { foo: number }) {
  const foo = props ?? 1

  return foo
}
```

> Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined.

**Comment by bradzacher** [comment]
I personally would argue that this doesn't really fit into this existing rule. Josh is correct that there is an implicit condition hidden in this operation - it desugars to `arg = arg === undefined ? default : arg` - but syntactically it is an assignment.

Personally I think that a better fit would be a type-aware extension rule for [`no-useless-assignment`](https://eslint.org/docs/latest/rules/no-useless-assignment).

But we don't _love_ extension rules - so perhaps instead it would be better to live as it's own rule that we can add to the recommended set?

**Comment by TkDodo** [comment]
yeah I don’t know enough about how you’d like to structure the rules; I’m fine with either 👍

**Comment by bradzacher** [comment]
Also worth noting that we could expand this to all destructuring and not just argument destructuring.

Eg the rule could also flag:
```ts
const {x = 1} = {x: 1};
const [y = 1] = [1];
```

**Comment by ronami** [comment]
+1 this would be useful though I think it's a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/6106.

**Comment by TkDodo** [comment]
> +1 this would be useful though I think it's a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/6106.

indeed; feel free to close as duplicate

**Comment by JoshuaKGoldberg** [comment]
D'oh, and the last commenter in 6106 was me, 22 months ago. Sorry for the switch-up TkDodo!
